src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.java

```diff
@@ -1552,42 +1550,65 @@ Collection<CompactionAggregate.UnifiedAggregate> getCompactionAggregates(Arena a
-                shardManager.splitSSTablesInShards(sstables,
-                                                   new ShardingStats(sstables, shardManager, controller).shardCountForDensity,
-                                                   (sstableShard, shardRange) -> Sets.newHashSet(sstableShard));
-
-                List<Set<CompactionSSTable>> oversizeGroups =
-                groups.stream()
-                      .filter(group -> group.size() > threshold * controller.getShardMaxSstablesFactor())
-                      .collect(Collectors.toList());
-
-                if (!oversizeGroups.isEmpty())
+                double shardThreshold = fanout * controller.getMaxSstablesPerShardFactor();
+                if (sstables.size() > shardThreshold)
                 {
-                    // Now combine the groups that share an sstable so that we have valid independent transactions.
-                    // Only keep the groups that were combined with an oversize group.
-                    groups = Overlaps.combineSetsWithCommonElement(groups);
-                    unbucketed.clear();
+                    List<Set<CompactionSSTable>> groups = shardManager.splitSSTablesInShards(sstables,
+                                                                                             controller.getNumShards(max),
+                                                                                             (sstableShard, shardRange) -> Sets.newHashSet(sstableShard));
 
-                    for (Set<CompactionSSTable> group : groups)
+
+                    List<Set<CompactionSSTable>> oversizeGroups = new ArrayList<>();
+                    for (Set<CompactionSSTable> compactionSSTables : groups)
                     {
-                        if (group.stream().anyMatch(
-                        sstable -> oversizeGroups.stream().anyMatch(oversizeGroup -> oversizeGroup.contains(sstable))))
+                        if (compactionSSTables.size() > shardThreshold)
                         {
-                            aggregates.add(
-                            CompactionAggregate.createUnified(group,
-                                                              Overlaps.maxOverlap(group,
-                                                                                  CompactionSSTable.startsAfter,
-                                                                                  CompactionSSTable.firstKeyComparator,
-                                                                                  CompactionSSTable.lastKeyComparator),
-                                                              createPick(controller, nextTimeUUID(), index, group),
-                                                              Collections.emptyList(),
-                                                              arena,
-                                                              this));
+                            oversizeGroups.add(compactionSSTables);
```

Comment:
Could you add a comment explaining why you prefer to save the groups and check them individually instead of adding all to a single set?
(So that someone reading it doesn't immediately want to do an optimization that may be counterproductive.)

Reply:
Well, this optimization would be good, wouldn't it ?
I think I should do it...